### ISSUE TITLE
[10.0]No more foreign keys in all tables when using this repoo

### ISSUE THREAD

**Issue by flotho** [OCA/pos#200]
Hi everyone,

This one seems really weird. First I suspected that Odoo CE has removed FK in fresh databases : https://github.com/odoo/odoo/issues/18880.
Yet starting with a fresh install with no specific addons in path make the things going well. Then repo by repo I added the dependencies and try to create new DB.
At last, this repo OCA/pos seeems to cause a creation with no Foreign keys.
I've updated the latest OCA/pos check all the depenecies, requirements etc, try many times to create with or without the dependencies. The result is the same, using OCA/pos cause no FK in the db!
Here are the logs, nothing special for me,any help would be appreciated.
[odoo.log.txt](https://github.com/OCA/pos/files/1232671/odoo.log.txt)

**Comment by flotho** [comment]
Hi,
I made additionnal tests :
* OCA/pos : 10.0 ddfab70 OCA Transbot updated translations from Transifex
* odoo/odoo : 10.0 89cade3bffc [FIX] pos_restaurant: wrap product names on OrderChangeReceipt

Same result, all the FK disappearing from tables, Does anyone reproduce this ?

**Comment by pedrobaeza** [comment]
Which module from OCA/pos? Check in runbot if it happens the same.

**Comment by flotho** [comment]
I'm trying each module one by one to identify

**Comment by flotho** [comment]
Well it looks like that pos_remove_category caused the problem. I'll recheck

**Comment by flotho** [comment]
Well, I confirm that having the module pos_remove_category cause the problem. I made those tests :
* Creating db with adding module from oca/pos one by one, retsart odoo between each addition
** => everything work fine except when pos_remove_category is present in the addons path
* Creating db with all the module except pos_remove_category
** => Everything Ok
* creating db with only pos_remove_categ => make FK disappear

i'm suspecting the ir.module.module dependency so that I tried to comment the init part.

There's something I can't log/ identify more without your support. Thanks in advance.
regards

**Comment by flotho** [comment]
@pedrobaeza it's not really easy for me to try on runbot because no error make the creation failed. Sadly there also nothing special in the logs. The db is created but the table structure haven't got anymore FK. I discovered that when a product.template has been deleted and its associated product.product still living its own ....
not easy to debug this one

**Comment by legalsylvain** [comment]
Thanks @flotho for reporting this problem.

This line seems to create your bug. (not tested)
https://github.com/OCA/pos/blob/10.0/pos_remove_pos_category/models/product.py#L89

Solution 1 : full refactoring
My point of view is that this module is not very clean, and should be refactored : 
- do not alter database.
- while loading product, remove pos_categ_id field, add the product_categ_id, and at the end of the load, initialize the pos_categ_id with the product_categ_id
- drop auto_end function and all product overload.
- product_category.available_in_pos : could be computed. (optional) (if category has products that can be sold in PoS.)

Solution 2 : more easy. 
- disable _auto_end function. (The module will not work anymore correctly), but the repo will be safe.

@simone, @sylvainc. Could you take a look ? 

Note : this problem is present on odoo8 too.
https://github.com/OCA/pos/blob/8.0/pos_remove_pos_category/product.py#L105

**Comment by flotho** [comment]
Hi everyone, 
Additionnal question, once this module is not present in the addons_path, do you think that an -u all will re-add all the FK on an already created database?
regards

**Comment by guewen** [comment]
Worth to note this module has another big design issue. It **replaces** the Manyone towards `pos.category` on the field `product_template.pos_categ_id` by a Many2one towards `product.category`. As soon as you will want to upgrade the base module `point_of_sale`, this one will try to create a Foreign Key towards `product.category` and fail (letting the database blocked in an intermediate state) because the ids might not exist in this table.

IMO this module should either be reworked in a clean way, either be removed.

**Comment by pedrobaeza** [comment]
Yeah, this is very critical. Who was the original author/migrator? Is he/she (or anyone else) going to attend this? If not, we should remove it.

**Comment by sylvainc** [comment]
Hello,

I'can only reply shortly sorry, i have not a easy fix because monkey
patching like this (in early startup) is too tricky and must be definitivly
avoid , please remove this module.
I'll try to find a workarround for the corrupted database.

Tanks for reporting

My mistake,

Regards,

Le 21 août 2017 9:53 AM, "Pedro M. Baeza" <notifications@github.com> a
écrit :

Yeah, this is very critical. Who was the original author/migrator? Is
he/she (or anyone else) going to attend this? If not, we should remove it.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/OCA/pos/issues/200#issuecomment-323672502>, or mute the
thread

**Comment by legalsylvain** [comment]
Done : PR to review : 
8.0 : #201 
9.0 : #202 
10.0 : #203

**Comment by flotho** [comment]
Thanks folks, needless to say that it had badside effects on my actual prod databases. Hopefully, -u all re-added all the FK.
regards

**Comment by simahawk** [comment]
@flotho thanks for spotting this! :+1: for removal.

**Comment by StefanRijnhart** [comment]
In case you were wondering like I was, this problem did not affect 8.0 or 9.0. The bug was introduced here, during the migration to 10.0, by not returning a call to super but returning only a reference to the super method instead: https://github.com/OCA/pos/pull/158/files#diff-a2bb6a42189587ba52241d71ea03f98eL104

Thank you @flotho for doing all this investigation and pinpointing the problem to this particular module.

**Comment by legalsylvain** [comment]
Thanks @StefanRijnhart for this explanation.

**Comment by simahawk** [comment]
holy :hankey: I'm gonna bury myself... :cry: 
sorry for the mess guys... :gun:

**Comment by Cedric-Pigeon** [comment]
@pedrobaeza @sylvainc @flotho @legalsylvain @simahawk Can you restore this addon. The fix is proposed for a long time and works fine... #182

**Comment by pedrobaeza** [comment]
Please restore the addon and add the patch in your PR, and if everything is correct, we merge at once.

**Comment by Cedric-Pigeon** [comment]
Please revert your merge, it is more simple. I pointed everyone in my PR months ago.

**Comment by pedrobaeza** [comment]
Sorry, but we can't allowed a module that breaks whole installation. Restore it in your PR and when merged, everything will work correctly. You can press yourself in revert commit, and GitHub will propose a new PR. Cherry-pick your commit in top of it and it's done.

**Comment by Cedric-Pigeon** [comment]
Ok you are right. I renamed my PR description and add the revert of the removal.

**Comment by pedrobaeza** [comment]
Thanks for the understanding, @Cedric-Pigeon

**Comment by sylvainc** [comment]
Thanks you all

**Comment by alexis-via** [comment]
I just discovered I am badly hit by this bug on a v10 database for a customer that went live on july 2017.

Running **-u all** doesn't restore the FK on the M2M relation tables, such as "res_partner_res_partner_category_rel" for example. On this specific v10 database created in july 2017:
```
test1=# \d res_partner_res_partner_category_rel
  Table "public.res_partner_res_partner_category_rel"
   Column    |  Type   | Collation | Nullable | Default 
-------------+---------+-----------+----------+---------
 category_id | integer |           | not null | 
 partner_id  | integer |           | not null | 
Indexes:
    "res_partner_res_partner_category_rel_category_id_partner_id_key" UNIQUE CONSTRAINT, btree (category_id, partner_id)
    "res_partner_res_partner_category_rel_category_id_idx" btree (category_id)
    "res_partner_res_partner_category_rel_partner_id_idx" btree (partner_id)
```

On a regular v10 database:
```
oldv10=# \d res_partner_res_partner_category_rel
  Table "public.res_partner_res_partner_category_rel"
   Column    |  Type   | Collation | Nullable | Default 
-------------+---------+-----------+----------+---------
 category_id | integer |           | not null | 
 partner_id  | integer |           | not null | 
Indexes:
    "res_partner_res_partner_category_rel_category_id_idx" btree (category_id)
    "res_partner_res_partner_category_rel_category_id_partner_id_key" UNIQUE CONSTRAINT, btree (category_id, partner_id)
    "res_partner_res_partner_category_rel_partner_id_idx" btree (partner_id)
Foreign-key constraints:
    "res_partner_res_partner_category_rel_category_id_fkey" FOREIGN KEY (category_id) REFERENCES res_partner_category(id) ON DELETE CASCADE
    "res_partner_res_partner_category_rel_partner_id_fkey" FOREIGN KEY (partner_id) REFERENCES res_partner(id) ON DELETE CASCADE
````

From my experience, this causes problems during openupgrade migration v13 to v14, when Odoo tries to add the FK on the M2M relation tables and fails to do so due to the entries in the relation table that have IDs that have been removed in the meantime in the regular table. This is a big issue on M2M relation tables that link to objects created via XMLIDs such as ir_ui_menu_group_rel, rule_group_rel, etc...

I think the good strategy is to restore the FK with ON DELETE CASCADE in the relation tables in v10 before starting openupgrade (because openupgrade will add/remove many entries due to changes in XMLIDs). If has a solution to restore the FK in the relation tables in v10... I'm very interested !

**Comment by alexis-via** [comment]
Here is a dirty patch for odoo v14 that cleans-up the relation table from dead IDs just before odoo tries to add the FK constraint:

https://pastecode.io/s/3p7v3xph

I hope it can be useful for those who were impacted by this horrible bug and have the same problem with missing FK in M2M relation tables and face problems during v13 to v14 openupgrade migration.
